### ISSUE TITLE
Questions and feedback and concerning the GLightbox API

### ISSUE THREAD

**Issue by acc987** [biati-digital/glightbox#309]
I am currently testing GLightbox as an alternative to the current light box script that I use. As part of my initial testing, I have created an extension that allows for a hash-based navigation to/in the light box instance. While getting familiar with GLightbox, I have accumulated some feedback concerning the current GLightbox API, i.e. the functions and events that are documented in your README, that I would like to share with you and I would like to discuss possible extensions of this API as I believe that I am currently querying internal variables of the GLightbox instance. I will begin by stating my questions and suggestions (independently of my extension), then I will motivate my suggestions by explaining how this would simplify my extension or avoid breaking changes, and finally I will share the source code of my extension so you are able to fully understand my motivation for these requests.

**Please describe the feature you want to be implemented**
My list of questions and suggestions is as follows:

1. Is it safe to invoke GLightbox API functions in an [IIFE](https://developer.mozilla.org/en-US/docs/Glossary/IIFE) that runs in a `<script>`-tag when referring to a GLightbox instance that was created before invoking the IIFE? From what I can tell using IIFEs is generally discouraged as this [apparently does not wait for the DOM to be computed](https://stackoverflow.com/a/31171096), but I am not sure whether this affects the behavior of GLightbox.
2. What's the recommended way to get the number of elements registered with a GLightbox instance? Given an instance named `lightbox`, I am currently querying `lightbox.elements.length`, but this is technically not part of the public API.
3. What's the recommended way to query whether a GLightbox instance has already been opened? Given an instance named `lightbox`, I am currently querying `lightbox.lightboxOpen`, but this is technically not part of the public API.
4. ~I have noticed that the `'slide_changed'` handler is invoked both when changing to another slide and when the GLightbox instance is initially opened. The `'slide_before_change'` handler, however, is only invoked when a slide is about to be changed. Would it be possible to also invoke this handler when opening the instance to match the `'slide_changed'` behavior?~ This might not necessarily be a good idea, see my follow-up comment below.
5. I have noticed that the `'open'` handler is invoked before the GLightbox instance actually opens and that the `'close'` handler is invoked only after the instance has fully closed (and the scrollbar has been restored). Would it be possible to change this such that the `'close'` handler is invoked before the closing animation begins, similar to the `'open'` handler? Alternatively, would you mind introducing a set of `'before_close'`/`'close'` API functions that are invoked before beginning with UI-related changes/after UI-related changes are complete? (This naming scheme would also allow `'close'` to maintain its current meaning, if this is important to you.)

**Explain why the feature is useful**
With numbers referring to the questions/suggestions above:

2. Having a documented API for querying the number of elements would provide a reliable way of checking index bounds before changing slides, see my example down below.
3. Having a documented API for determining whether an GLightbox instance has been opened would improve reliability of code that needs to query this state, see my example down below.
4. ~If the `'slide_before_change'` handler is also invoked when opening the GLightbox instance, this would allow me to get rid of my `'open'` handler, which currently relies on `getActiveSlideIndex()` being set correctly when invoking the handler while opening the GLightbox instance and feels a bit hacky, see my example down below. Note that using a `'slide_changed'` handler instead would suffer from a race condition when the user decides to reload the page after having clicked on an element that is registered with GLightbox but before the slide change is complete, in the sense that the selected image would not be re-opened in GLightbox as the hash was not updated in time.~
5. With the current behavior, the hash of the URL is only updated after the GLightbox instance has fully closed. This is unfortunate, as the user might request closing the instance and refresh the page, in which case the GLightbox instance would be re-opened as the hash was not yet updated in this case. (I currently see no existing way to hook into the closing-requested stage of GLightbox.)

**Additional context**
DISCLAIMER: I am a JavaScript novice and while I am happy to receive feedback for my source code from you when answering my questions and replying to my suggestions, please keep in mind that the source code is mainly provided as a working example to motivate my questions/suggestions. Note in particular, that the extension should be considered a work in progress (as you can tell from my many edits of this post...).

Unfortunately, given the current implementation of the GLightbox API, the source code suffers from various race conditions, see my follow-up comment below, so I would discourage anyone from actually using this extension.

The source code for my extension, located in a `<script>`-tag in the body of my HTML document that also loads GLightbox, is as follows:
```JavaScript
const gallery = 'glightbox';
var lightbox = GLightbox({
  selector: '.' + gallery
});
// Allow for persistent links to specific images by using the hash of the URL:
lightbox.on('slide_before_change', ({previous, current}) => {
  const {slideIndex, ...remainder} = current;
  const desiredHash = '#' + gallery + '-' + (slideIndex + 1);
  if (window.location.hash !== desiredHash) {
    window.location.hash = desiredHash;
  }
});
lightbox.on('open', () => {
  const desiredHash = '#' + gallery + '-' + (lightbox.getActiveSlideIndex() + 1);
  if (window.location.hash !== desiredHash) {
    window.location.hash = desiredHash;
  }
});
lightbox.on('close', () => {
  if (window.location.hash !== '') {
    // Prevent scrolling to the top-left by restoring the old offsets after changing the hash,
    // motivated by https://stackoverflow.com/a/5298684 and https://stackoverflow.com/a/39657473.
    var scrollElem = document.scrollingElement || document.documentElement;
    const scrollTopOld = scrollElem.scrollTop;
    const scrollLeftOld = scrollElem.scrollLeft;
    window.location.hash = '';
    scrollElem.scrollTop = scrollTopOld;
    scrollElem.scrollLeft = scrollLeftOld;
  }
});
// Utility function for extracting light box data from the hash of the URL:
function hash2index() {
  const hashIsImageIdentifier = window.location.hash.match(/^#(.*)-(\d+)$/);
  if (hashIsImageIdentifier) {
    const hashAlbum = hashIsImageIdentifier[1];
    const hashIndex = parseInt(hashIsImageIdentifier[2], 10) - 1;
    if (hashAlbum === gallery && hashIndex >= 0) {
      return hashIndex;
    }
  }
  return NaN;
}
// Update the light box when the hash of the URL is changed manually:
window.addEventListener('hashchange', function() {
  const hashIndex = hash2index();
  if (hashIndex < lightbox.elements.length) {
    if (lightbox.lightboxOpen) {
      const activeIndex = lightbox.getActiveSlideIndex();
      if (hashIndex != activeIndex) {
        lightbox.goToSlide(hashIndex);
      }
    } else {
      lightbox.openAt(hashIndex);
    }
  } else if (window.location.hash === '' && lightbox.lightboxOpen) {
    lightbox.close();
  }
});
// Open the light box immediately if an image is already specified when loading this page:
(function () {
  const hashIndex = hash2index();
  if (hashIndex < lightbox.elements.length) {
    lightbox.openAt(hashIndex);
  }
})();
```

**Comment by acc987** [comment]
An amendment concerning my fourth question/suggestion: Being able to distinguish between the current slide changing due to the lightbox (re-)openening and the current slide changing when already in the lightbox is of course useful, and when following my suggestion would probably require further changes to distinguish between the two events in the `slide_before_change` handler. Leaving the implementation of the API as is, would already offer such a distinction, so I don't mind if you ignore this suggestion (as long as it is safe to call `getActiveSlideIndex()` in the `'open'` handler).

As for the use of the API functions in my above example, I have unfortunately realized that the above code suffers from various race conditions as it seems that the event handler spawns threads that execute concurrently and a GLightbox instance `lightbox` breaks when `lightbox.openAt`/`lightbox.goToSlide` and `close` calls overlap. This can happen when quickly navigating through the browser's history, which is updated due to the `window.location.hash` changes. Race conditions can most likely also occur when history-based and mouse/keyboard-based navigation occur simultaneously. From what I can tell, one would need to make changes to GLightbox itself to allow for a serialization of parallel interactions or "abort" previously API calls when the next one is received, in particular since mouse/keyboard-based navigation belongs to the internals of GLightbox.

If a future version of GLightbox would allow for overlapping API calls (without breaking the internal state of the instance) that would of course be amazing, but I guess until then one cannot provide a hash-based navigation that uses the browser's history and reacts on manual changes to the URL without introducing race conditions. Even then, one would also need to be able to tell a GLightbox instance not to invoke the corresponding event handler when an API call is invoked as updating the URL is not necessary when the user is navigating through the history. Note that one way to implement this, is to allow passing an optional object to GLightbox API calls, which is then passed on to the event handlers that have been registered. The event handlers can then use the information provided by this object to, e.g., decide whether or not to change `window.location.hash` or use the History API (i.e. `history.pushState`/`history.replaceState`) to make changes to the current URL.

I hope that my example demonstrates that realizing these two additional suggestions would make GLightbox a lot more versatile/customizable, but I fear that they would require extensive changes to its internals, so I am not getting my hopes up...

**Comment by Andrew-web-coder** [comment]
I think you're just wasting your time, author said it works on v4 and it will be built completely differently. So I would suggest to stick with alternatives that support this feature out of the box, like https://fancyapps.com/docs/ui/fancybox/

**Comment by frankie-tech** [comment]
@acc987 Thanks for the in depth questions! I'm working on updating the documentation, so this helps me a lot!

I'll go in order and see if we can hash it out:

> 1. Is it safe to invoke GLightbox API functions in an IIFE that runs in a <script>-tag when referring to a GLightbox instance that was created before invoking the IIFE? From what I can tell using IIFEs is generally discouraged as this apparently does not wait for the DOM to be computed, but I am not sure whether this affects the behavior of GLightbox.

You probably answered your own question here. You don't get any benefit from running GLightbox inside of an IIFE over a ready/DOMContentLoaded event listener over a script tag with the defer attribute on it. 

As long as GLightbox runs *after* all of your slide elements are in the DOM, everything should work as expected. If you add more slide elements to the DOM, you will need to make the GLightbox again.

> 2. What's the recommended way to get the number of elements registered with a GLightbox instance? Given an instance named lightbox, I am currently querying lightbox.elements.length, but this is technically not part of the public API.

No, you've got it right! One of the ways you can create a GLightbox is using an array of elements, [docs](https://github.com/biati-digital/glightbox#lightbox-options). GLightbox instances use the elements property internally whether you use a selector or an elements array. You could consider the array a more "*advanced*" way of doing it, but it's just a way to create/add slides without adding to the DOM. 

> 3. What's the recommended way to query whether a GLightbox instance has already been opened? Given an instance named lightbox, I am currently querying lightbox.lightboxOpen, but this is technically not part of the public API.

I'm reading this two ways, so I'll go over both real quick.

1. Let's say you are in a scenario where you have 2 lightboxes on a page, `lightboxA` and `lightboxB`, how can you tell if any of them are open, or which of the two is open:

You are more than welcome to use `lightboxA.lightboxOpen` to see if it is open or not. Alternatively, there have been people who have created small hooks to put in their own solutions. It's encouraged to read through the code, log the GLightbox instance, and see what is at your disposal.

That being said, I understand that when writing *Enterprise* level code developers don't have that luxury. While `lightbox.lightboxOpen` is not written down in the documentation as a public API, I would say it is safe to use for v3.

2. Let's say you are in a scenario where you have a lightbox `lightboxOne`, and you need to know if it was previously opened. This is due to code needing to be run the first time it is open, but not any of the following times.

There isn't a way to track this currently, but GLightbox instances have a `once` method for adding one-time event listeners.

```js
lightboxOne.once('open', () => { /* your function here */ });
```

> 5. I have noticed that the `open` handler is invoked before the GLightbox instance actually opens and that the 'close' handler is invoked only after the instance has fully closed (and the scrollbar has been restored). Would it be possible to change this such that the 'close' handler is invoked before the closing animation begins, similar to the 'open' handler? Alternatively, would you mind introducing a set of `before_close`/`close` API functions that are invoked before beginning with UI-related changes/after UI-related changes are complete? (This naming scheme would also allow 'close' to maintain its current meaning, if this is important to you.)

Considering the other event names, you have hit the nail on the head with a `before_close`/`close` dichotomy. I imagine it wouldn't be hard to add in, and would be a good first commit. I will add it as a separate issue as I'd like to preserve this for my documentation notes.

> I have unfortunately realized that the above code suffers from various race conditions as it seems that the event handler spawns threads that execute concurrently and a GLightbox instance lightbox breaks when `lightbox.openAt`,`lightbox.goToSlide`, and `lightbox.close` calls overlap. This can happen when quickly navigating through the browser's history, which is updated due to the `window.location.hash changes`.

Ultimately, it sounds like what you're looking to do is persist the state of the lightbox across certain contexts, like a page refresh. Unfortunately that means you've now entered the realm of a thousand solutions and no answers.

GLightbox wasn't made with that in mind, and "shim"-ing it in is going to be a headache (as you have been feeling). I can't think of a way to easily fit that in without either unraveling/refactoring large parts of the GLightbox code.

Something that popped into my head is the idea of using your own event emitter as a sort of proxy to the GLightbox events/methods. This could give you a little bit more control over when code is running and could give you the opportunity to deal with the race conditions you're running into.

My hesitation (if I was you) is that it adds a layer of complexity that now is your responsibility to manage. 

> If a future version of GLightbox would allow for concurrent API calls (without breaking the internal state of the instance) that would of course be amazing...

I think you're identifying all of the reasons that it is tough to convert to/add a concurrency model to GLightbox perfectly. You've identified a use case for at least a *more* concurrent api really well.

I do wonder where this kind of addition/change falls in balancing GLightbox's ability to be a "drop in and get a lightbox" and a "extentable, developer friendly, it-is-what-you-make-it" type of package. Without knowing what is on the horizon for v4, it's hard to say. It does sound like a fit, though.

Let me know if there's something I didn't address or you'd like to talk about further.

**Comment by acc987** [comment]
Sorry for the late reply. I pretty much agree with your observations and appreciate your comments. Currently, I can accept the limitations of my extension as it is for personal use. I have already read elsewhere on GitHub that version 4 will feature extensive changes, so I will certainly (have to) re-evaluate my code when that version is (eventually) released. If the internal implementation is more friendly towards my particular use case that would of course be nice, but I understand that many other things have to be taken into consideration and that the scope of version 4 is not yet determined.
